### PR TITLE
Miri rename undef uninit

### PR DESCRIPTION
**Pull request by pnadon**
Renamed parts of code within the `librustc_middle/mir/interpret/` directory.

Related issue [#71193](https://github.com/rust-lang/rust/issues/71193)

### CODE REVIEW

**Comment by lcnr** [comment]
:+1:  r? @oli-obk cc @RalfJung

File: src/librustc_middle/mir/interpret/allocation.rs [hunk 1]
```diff
@@ -565,7 +565,7 @@ impl<'tcx, Tag: Copy, Extra> Allocation<Tag, Extra> {
         })
     }
 
-    pub fn mark_definedness(&mut self, ptr: Pointer<Tag>, size: Size, new_state: bool) {
+    pub fn mark_init(&mut self, ptr: Pointer<Tag>, size: Size, new_state: bool) {
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/allocation.rs:568]: `new_state` is not very clear; maybe `is_init` or so?

**Comment by RalfJung** [comment]
This looks great apart from one nit. :)

File: src/librustc_middle/mir/interpret/allocation.rs (hunk 1 above)
Comment (pnadon) [reply src/librustc_middle/mir/interpret/allocation.rs:568]: The latter does make more sense.

**Comment by RalfJung** [comment]
Looking great, thanks. :)
r? @RalfJung @bors r+

@pnadon do you think you will have time to fix https://github.com/rust-lang/miri/ as well? Let me know if you need any help with that.

**Comment by pnadon** [comment]
> Looking great, thanks. :)
> r? @RalfJung @bors r+
> 
> @pnadon do you think you will have time to fix https://github.com/rust-lang/miri/ as well? Let me know if you need any help with that.

Do you mean the changes referenced in [this comment](https://github.com/rust-lang/rust/issues/71193#issuecomment-662531351)? If so I would be glad to!

**Comment by pnadon** [comment]
> @pnadon do you think you will have time to fix https://github.com/rust-lang/miri/ as well? Let me know if you need any help with that.

I assume you may have been referring to the renames in `src/librustc_mir/interpret`, in which case I would like to get to those as well, but I think I will leave them unclaimed so that someone else has a chance to work on them while I continue working within `src/librustc_middle/mir/interpret`.

**Comment by RalfJung** [comment]
> Do you mean the changes referenced in this comment? If so I would be glad to!

No I mean that Miri (the standalone tool at https://github.com/rust-lang/miri/, not the Miri core engine which this PR touches) is using some of the APIs you renamed here, and will fail to build once this PR lands. Someone has to get it to build again. :)

**Comment by pnadon** [comment]
> > Do you mean the changes referenced in this comment? If so I would be glad to!
> 
> No I mean that Miri (the standalone tool at https://github.com/rust-lang/miri/, not the Miri core engine which this PR touches) is using some of the APIs you renamed here, and will fail to build once this PR lands. Someone has to get it to build again. :)

Oh I see, yes I'll look into that!

**Comment by RalfJung** [comment]
As for the remaining renames mentioned in https://github.com/rust-lang/rust/issues/71193, that is an orthogonal question, feel free to continue working on those as well. :) Beyond what is listed in https://github.com/rust-lang/rust/issues/71193#issuecomment-626128359, I have not yet taken stock what even needs to be done.

**Comment by pnadon** [comment]
Actually, would you want me to finish the remaining renames in `src/librustc_middle/mir/interpret`, and then start working on [https://github.com/rust-lang/miri/](https://github.com/rust-lang/miri/)? That way there is less back-and-forth.

**Comment by RalfJung** [comment]
Miri should always build, so once this lands a PR needs to be prepared ASAP. It would be even better if you could prepare that PR before this one here lands. :D

**Comment by pnadon** [comment]
> Miri should always build, so once this lands a Miri PR needs to be prepared ASAP. It would be even better if you could prepare that Miri PR before this one here lands. :D

That makes sense, since this morning I have also renamed the 3 other functions / structs mentioned in that comment, I guess what I was trying to say was that I could commit them here, and then start working on https://github.com/rust-lang/miri/, once I get the changes made in the Miri repo, I could open a PR there and then we could close both simultaneously. Does that sound good?

**Comment by RalfJung** [comment]
Oh sure, we can add more renames to this PR.
@bors r-

**Comment by pnadon** [comment]
I ripgrep'd through Miri's src and only found ScalarMaybeUninit::not_undef, as well as a function call `Immediate::ScalarPair(old.to_scalar_or_undef()`. The former I renamed, the latter was found within  `Immediate::to_scalar_or_undef` in `src/librustc_mir/interpret/allocation.rs`

**Comment by pnadon** [comment]
i also ripgrep'd the other changes I just made in Miri, and I could not find them. This seems strange to me, but I will open a PR on Miri with what I have so far and continue my work there.

**Comment by RalfJung** [comment]
The instructions at https://github.com/rust-lang/miri/blob/master/CONTRIBUTING.md#building-miri-with-a-locally-built-rustc explain how you can build-test Miri against your locally built rustc. Make sure you use a recent rustc master though to avoid build failures due to unrelated changes.

File: src/librustc_middle/mir/interpret/allocation.rs [hunk 2]
```diff
@@ -573,28 +573,28 @@ impl<'tcx, Tag: Copy, Extra> Allocation<Tag, Extra> {
     }
 }
 
-/// Run-length encoding of the undef mask.
+/// Run-length encoding of the uninit mask.
 /// Used to copy parts of a mask multiple times to another allocation.
-pub struct AllocationDefinedness {
+pub struct InitMaskCompressed {
     /// The definedness of the first range.
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/allocation.rs:579]:
```suggestion
    /// Whether the first range is initialized.
```

File: src/librustc_middle/mir/interpret/allocation.rs [hunk 3]
```diff
@@ -573,28 +573,28 @@ impl<'tcx, Tag: Copy, Extra> Allocation<Tag, Extra> {
...
     initial: bool,
     /// The lengths of ranges that are run-length encoded.
     /// The definedness of the ranges alternate starting with `initial`.
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/allocation.rs:582]:
```suggestion
    /// The initialization state of the ranges alternate starting with `initial`.
```

File: src/librustc_middle/mir/interpret/allocation.rs [hunk 4]
```diff
@@ -573,28 +573,28 @@ impl<'tcx, Tag: Copy, Extra> Allocation<Tag, Extra> {
...
     ranges: smallvec::SmallVec<[u64; 1]>,
 }
 
-impl AllocationDefinedness {
-    pub fn all_bytes_undef(&self) -> bool {
+impl InitMaskCompressed {
+    pub fn no_bytes_init(&self) -> bool {
         // The `ranges` are run-length encoded and of alternating definedness.
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/allocation.rs:588]:
```suggestion
        // The `ranges` are run-length encoded and of alternating initialization state.
```

File: src/librustc_middle/mir/interpret/allocation.rs [hunk 5]
```diff
@@ -573,28 +573,28 @@ impl<'tcx, Tag: Copy, Extra> Allocation<Tag, Extra> {
...
-        // So if `ranges.len() > 1` then the second block is a range of defined.
+        // So if `ranges.len() > 1` then the second block is an initialized range.
         !self.initial && self.ranges.len() == 1
     }
 }
 
 /// Transferring the definedness mask to other allocations.
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/allocation.rs:594]:
```suggestion
/// Transferring the initialization mask to other allocations.
```
I am stopping here; please also replace the remaining uses of "definedness" if there are any.

File: src/librustc_middle/mir/interpret/allocation.rs [hunk 6]
```diff
@@ -625,13 +625,13 @@ impl<Tag, Extra> Allocation<Tag, Extra> {
 
         ranges.push(cur_len);
 
-        AllocationDefinedness { ranges, initial }
+        InitMaskCompressed { ranges, initial }
     }
 
     /// Applies multiple instances of the run-length encoding to the undef mask.
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/allocation.rs:631]:
```suggestion
    /// Applies multiple instances of the run-length encoding to the initialization mask.
```

File: src/librustc_middle/mir/interpret/allocation.rs [hunk 7]
```diff
@@ -573,28 +573,28 @@ impl<'tcx, Tag: Copy, Extra> Allocation<Tag, Extra> {
...
     }
 }
 
 /// Transferring the definedness mask to other allocations.
 impl<Tag, Extra> Allocation<Tag, Extra> {
     /// Creates a run-length encoding of the undef mask.
-    pub fn compress_undef_range(&self, src: Pointer<Tag>, size: Size) -> AllocationDefinedness {
+    pub fn compress_undef_range(&self, src: Pointer<Tag>, size: Size) -> InitMaskCompressed {
         // Since we are copying `size` bytes from `src` to `dest + i * size` (`for i in 0..repeat`),
         // a naive undef mask copying algorithm would repeatedly have to read the undef mask from
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/allocation.rs:599]:
```suggestion
        // a naive initialization mask copying algorithm would repeatedly have to read the initialization mask from
```

File: src/librustc_mir/interpret/memory.rs [hunk 8]
```diff
@@ -907,7 +907,7 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'mir, 'tcx, M> {
         // Prepare a copy of the undef mask.
         let compressed = self.get_raw(src.alloc_id)?.compress_undef_range(src, size);
 
-        if compressed.all_bytes_undef() {
+        if compressed.no_bytes_init() {
             // Fast path: If all bytes are `undef` then there is nothing to copy. The target range
```
Comment (RalfJung) [inline src/librustc_mir/interpret/memory.rs:911]: More "undef" here; please also replace those.

**Comment by pnadon** [comment]
Ok, aside from the comments though I'm assuming everything looks ok. In that case, I will work on getting Miri to build, and also fix the above comments.

**Comment by pnadon** [comment]
Miri seems to build correctly, and I have changed the above comments.

File: src/librustc_middle/mir/interpret/allocation.rs [hunk 9]
```diff
@@ -576,27 +576,27 @@ impl<'tcx, Tag: Copy, Extra> Allocation<Tag, Extra> {
 /// Run-length encoding of the uninit mask.
 /// Used to copy parts of a mask multiple times to another allocation.
 pub struct InitMaskCompressed {
-    /// The definedness of the first range.
+    /// Whether the first range is initialized.
     initial: bool,
     /// The lengths of ranges that are run-length encoded.
-    /// The definedness of the ranges alternate starting with `initial`.
+    /// The initialization state of the ranges alternate starting with `initial`.
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/allocation.rs:582]:
```suggestion
    /// The initialization state of the ranges alternates starting with `initial`.
```

File: src/librustc_middle/mir/interpret/allocation.rs [hunk 10]
```diff
@@ -576,27 +576,27 @@ impl<'tcx, Tag: Copy, Extra> Allocation<Tag, Extra> {
...
+        // The `ranges` are run-length encoded and of alternating initialization state.
         // So if `ranges.len() > 1` then the second block is an initialized range.
         !self.initial && self.ranges.len() == 1
     }
 }
 
-/// Transferring the definedness mask to other allocations.
+/// Transferring the initialization mask to other allocations.
 impl<Tag, Extra> Allocation<Tag, Extra> {
     /// Creates a run-length encoding of the undef mask.
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/allocation.rs:596]: Still an "undef" left here.

File: src/librustc_middle/mir/interpret/allocation.rs [hunk 11]
```diff
@@ -544,18 +544,18 @@ impl<'tcx, Tag: Copy, Extra> Allocation<Tag, Extra> {
 
 /// Undefined bytes.
 impl<'tcx, Tag: Copy, Extra> Allocation<Tag, Extra> {
-    /// Checks whether the given range  is entirely defined.
+    /// Checks whether the given range  is entirely initialized.
     ///
-    /// Returns `Ok(())` if it's defined. Otherwise returns the range of byte
+    /// Returns `Ok(())` if it's initialized. Otherwise returns the range of byte
     /// indexes of the first contiguous undefined access.
```
Comment (RalfJung) [inline src/librustc_middle/mir/interpret/allocation.rs:550]: Another remaining "undef".

Comment (pnadon) [reply src/librustc_middle/mir/interpret/allocation.rs]: I searched again through allocation.rs and found this comment as well as one other and a variable, which were all changed appropriately. I left comments which were for other functions / parts of the code which have yet to be changed (eg. `inspect_with_undef_and_ptr_outside_interpreter`).

**Comment by RalfJung** [comment]
There are still two unresolved comments left (you can see them by scrolling through https://github.com/rust-lang/rust/pull/74664/files)

**Comment by pnadon** [comment]
> There are still two unresolved comments left (you can see them by scrolling through https://github.com/rust-lang/rust/pull/74664/files)

I had left some of them since they were related to functions which still had "undef" within their name (and are currently being discussed in the related issue) to try and group the changes in the comments with the changes in the actual code, however since many of them still use the functions which have been changed, I believe it might be best to do the rest of the comments. in allocation.rs anyways.

As for renaming the rest of the functions, I think I will hold off on that until we have decided on the new names in the issue, and do them all at once for both the Rust and Miri respositories.

**Comment by RalfJung** [comment]
Yes this is great. :) Could you squash the commits into one (or a few; 12 seems a bit too much)? Then I'll r+.

**Comment by pnadon** [comment]
Squashed into 2 commits

**Comment by RalfJung** [comment]
Thanks. :) @bors r+

**Comment by pnadon** [comment]
Thank you @RalfJung for your patience and help throughout this, this was my first contribution to Rust and I really appreciated the time and effort you put into making sure I did things correctly.

**Comment by RalfJung** [comment]
@pnadon thanks for contributing, here's hoping that you'll continue to enjoy working on Rust. :)

**Pull request merged**